source/extensions/filters/http/geoip/geoip_filter.cc

```diff
@@ -0,0 +1,157 @@
+  lookupGeolocationAnonData(config_->geoAnonHostingHeader(), remote_address,
+                            &Driver::getIsAnonymousHostingProvider);
+  lookupGeolocationAnonData(config_->geoAnonTorHeader(), remote_address,
+                            &Driver::getIsAnonymousTorExitNode);
+  lookupGeolocationAnonData(config_->geoAnonProxyHeader(), remote_address,
+                            &Driver::getIsAnonymousPublicProxy);
+  all_lookups_invoked_ = true;
+
+  // Stop the iteration for headers for the current filter and the filters following.
+  return allLookupsComplete() ? Http::FilterHeadersStatus::Continue
+                              : Http::FilterHeadersStatus::StopIteration;
+}
+
+void GeoipFilter::lookupGeolocationData(
+    const absl::optional<std::string>& geo_header,
+    const Network::Address::InstanceConstSharedPtr& remote_address,
+    void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
+                         const LookupCallbacks& callbacks,
+                         const absl::optional<std::string>& geo_header) const) {
+  if (geo_header) {
+    ++num_lookups_to_perform_;
+    ((*driver_).*func)(remote_address, *this, geo_header);
+  }
+}
+
+void GeoipFilter::lookupGeolocationAnonData(
+    const absl::optional<std::string>& geo_header,
+    const Network::Address::InstanceConstSharedPtr& remote_address,
+    void (Driver::*func)(const Network::Address::InstanceConstSharedPtr& address,
+                         const LookupCallbacks& callbacks,
+                         const absl::optional<std::string>& geo_header) const) {
+  if (geo_header) {
+    ++num_lookups_to_perform_;
+    ((*driver_).*func)(remote_address, *this, geo_header);
+  }
+}
+
+Http::FilterDataStatus GeoipFilter::decodeData(Buffer::Instance&, bool) {
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus GeoipFilter::decodeTrailers(Http::RequestTrailerMap&) {
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void GeoipFilter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+void GeoipFilter::onLookupComplete(const absl::optional<std::string>& lookup_result,
+                                   const absl::optional<std::string>& geo_header) {
+  if (!request_headers_) {
+    ENVOY_LOG(debug, "Geoip filter: no header to decorate in reguest");
+    return;
+  }
+  ASSERT(num_lookups_to_perform_ > 0);
+  --num_lookups_to_perform_;
+  if (lookup_result) {
+    request_headers_->setCopy(Http::LowerCaseString(geo_header.value()), lookup_result.value());
+    config_->incHit(geo_header.value());
+  }
+  config_->incTotal(geo_header.value());
+  if (allLookupsComplete()) {
+    ENVOY_LOG(debug, "Geoip filter: finished decoding geolocation headers");
+    decoder_callbacks_->continueDecoding();
+  }
+}
+
+bool GeoipFilter::allLookupsComplete() {
+  return all_lookups_invoked_ && num_lookups_to_perform_ == 0;
```

Comment:
For now i think those two flags are enough to track the state of lookups completion per request. But i may have missed some edge cases